File: sudoku.py
# Time Complexity :
# O(N^2)  - Size of the Matrix

# Space Complexity :
# O(N)

# Did this code successfully run on Leetcode :
#Yes

#We do a row test, column test and grid test
#In row test, we go through the all the rows and check if all elements(not "." elements) are unique. If they aren't we return False
#We do the same in column test
#For thew grid test, we use the same logic of going through each grid, and each element in each grid and checking if they are all unique. We move been grids using 4 loop of  Sqrt(N) each

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        values = {}

        #IS valid row 
        for i in range(0,len(board)):
            for j in range(0,len(board[0])):
                if board[i][j] == '.':
                    continue
                else:
                    if board[i][j] in values :
                        return False
                    else :
                        values[board[i][j]] = 0

            values = {}

        #Is valid column 
        for j in range(0,len(board)):
            for i in range(0,len(board[0])):
                if board[i][j] == '.':
                    continue
                else:
                    if board[i][j] in values :
                        return False
                    else :
                        values[board[i][j]] = 0
            values = {}

        #is valid grid
        for i in range(0,3):
            for j in range(0,3):
                for k in range(i*3,(i*3)+3):
                    for l in range(j*3,(j*3)+3):
                        if board[k][l] == '.':
                            continue
                        else:
                            if board[k][l] in values :
                                return False
                            else :
                                values[board[k][l]] = 0
                values = {}
        return True

File: Sudoku.cpp
//Time Complexity: Constant Time
//Space Complexity: Constant
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for(int i = 0; i < 9; i ++) {
            for(int j = 0; j < 9; j++) {
                if(board[i][j] != '.') {
                    if(repeatInRow(board, i, j) || repeatInCol(board, i , j) || repeatSubCell(board, i, j)) {
                        return false; 
                    }
                }
                // Check if board[i][j] repeats in the row 

                // Check if board[i][j] repeats in the column 

                // Check if board[i][j] repeats in the sub-cell
            }
        }

        return true;        
    }


    bool repeatInRow(vector<vector<char>>& board, int row, int col) {
        int number = board[row][col];

        for(int j = 0; j < 9; j++) {
            if(j != col) {
                if(board[row][j] == number) {
                    cout << "return true for: " << row << " : " << j << endl; 
                    return true; 
                } 
            }
        }
        
        return false;     
    }


    bool repeatInCol(vector<vector<char>>& board, int row, int col) {
        int number = board[row][col];

        for(int i = 0; i < 9; i++) {
            if(i != row) {
                if(board[i][col] == number) {
                    cout << "return true for: " << i << " : " << col << endl; 
                    return true; 
                }
            }
        }

        return false;     
    }


    bool repeatSubCell(vector<vector<char>>& board, int row, int col) {
        int number = board[row][col];
        int startRow, startCol; 

        if(row >= 0 && row <= 2) {
            startRow = 0; 
        } else if(row >= 3 && row <= 5) {
            startRow = 3; 
        } else {
            startRow = 6;
        }

        if(col >= 0 && col <= 2) {
            startCol = 0; 
        } else if(col >= 3 && col <= 5) {
            startCol = 3; 
        } else {
            startCol = 6;
        }

        for(int i = startRow; i < startRow + 3; i++) {
            for(int j = startCol; j < startCol + 3; j++) {
                if(i != row && j != col) {
                    if(board[i][j] == number) {
                        cout << "return true for: " << i << " : " << j << endl; 
                        return true; 
                    }
                }
            }
        }

        return false; 
             
    }
};